Gnc

lucacarlone opened the pull request:
Added Graduated Non-Convexity (GNC) as a nonlinear optimizer in GTSAM. This is an implementation of the GNC paper (https://arxiv.org/pdf/1909.08605.pdf), but it uses iterative solvers in the inner loop instead of non-minimal global solvers.

Outstanding tasks:
- [x] reintroduce “using BaseOptimizer = BaseOptimizerParameters::OptimizerType;”  to let the user choose the iterative solver for the inner loop.
- [x] move GncOptimizer and its params to a proper class/file when ready.
- [x] populate the TLS option (now the code only implements GM = Geman McClure).
- [x] double-check that the 0.5 in front of the factors in gtsam does not require a modification to the weight update rule.

dellaert commented on tests/testGncOptimizer.cpp:
```diff
@@ -0,0 +1,625 @@
+  // Note: in difficult instances, we set the odometry measurements to be inliers,
+  // but this problem is simple enought to succeed even without that assumption
+  // std::vector<size_t> knownInliers;
+  GncParams<GaussNewtonParams> gncParams = GncParams<GaussNewtonParams>();
```
Just say `GncParams<GaussNewtonParams> gncParams;` ?

lucacarlone replied:
solved

dellaert commented on tests/testGncOptimizer.cpp:
```diff
@@ -0,0 +1,625 @@
+using symbol_shorthand::L;
+static double tol = 1e-7;
+
+/* ************************************************************************* */
```
Probably need to move to a .h file now

dellaert commented on tests/testGncOptimizer.cpp:
```diff
@@ -0,0 +1,625 @@
+  VerbosityGNC verbosityGNC = SILENT; /* verbosity level */
+  std::vector<size_t> knownInliers = std::vector<size_t>(); /* slots in the factor graph corresponding to measurements that we know are inliers */
+
+  void setLossType(const RobustLossType type) {
```
add comments to all methods, at least `/// one-liner doxygen`

dellaert commented on tests/testGncOptimizer.cpp:
```diff
@@ -0,0 +1,625 @@
+    GM /*Geman McClure*/, TLS /*Truncated least squares*/
+  };
+
+  using BaseOptimizer = GaussNewtonOptimizer; // BaseOptimizerParameters::OptimizerType;
```
Oh ?

lucacarlone replied:
yes, this exactly the problem we need to solve. I can do:
using BaseOptimizer = BaseOptimizerParameters::OptimizerType;
Then I can go to LevenbergMarquardtParams and define OptimizerType as LevenbergMarquardtOptimizer.
However, LevenbergMarquardtParams does not know about LevenbergMarquardtOptimizer, so I'm not sure how to make that define work. did you have something different in mind?

dellaert replied:
I think you need to modify LevenbergMarquardtParams to add the typedef there

lucacarlone replied:
ok I'll try that, but I suspect an easier way would be to template on the Optimizer instead of the Params.

dellaert reviewed (commented):
I left some comments. What is the main templating issue you wanted me to look at ?

lucacarlone commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,381 @@
+      // initialize mu to the value specified in Remark 5 in GNC paper
+      // degenerate case: residual is close to zero so mu approximately equals
+      // to -1
+      return params_.barcSq / (2 * rmax_sq - params_.barcSq);
```
@jingnanshi : I don't think mu can be negative. does it make sense to have a lower bound for mu in this and the GM case? it the max residual is below barcsq, it means there is no outlier and maybe we can just "return"

jingnanshi replied:
@lucacarlone I completely agree. I put the check outside the initializeMu function (see line 210 to 217) so that we can easily just return the current result.

lucacarlone replied:
great, @jingnanshi !

lucacarlone commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,381 @@
+    case GncParameters::GM:
+      return std::fabs(mu - 1.0) < 1e-9; // mu=1 recovers the original GM function
+    case GncParameters::TLS:
+      return std::fabs(mu - mu_prev) < params_.relativeMuTol;
```
@jingnanshi : this does not make much sense, since the decrease in mu is constant. maybe @pantonante has a better check?

jingnanshi replied:
@lucacarlone : good catch! this should be cost - cost_prev

lucacarlone replied:
I think Pasquale checks if the weights are (close to) binary, which is the case after TLS converges.

jingnanshi replied:
@lucacarlone : I just checked Pasquale's email, and he suggested using the relative change of the weighted sum of the squared residuals, which is what I used in TEASER++. I think I'll implement a checkConvergence function that checks mu or relative cost change depending on loss type.

pantonante replied:
Yes, I suggest to use the relative change of the weighted sum of the squared residuals because it works well in practice. Here seems you are checking that they are not changing much, this doesn't work for TLS.

lucacarlone replied:
@jingnanshi : I would implement this in the main loop (before "if (checkMuConvergence(mu, mu_prev))"): I would simply add "if (checkCostConvergence(..)  OR checkWeightsConvergence(..))" and implement these in a separate function. Let me know if you want me to take a stab - I can try to do that tonight

jingnanshi replied:
@lucacarlone : That makes sense. I won't be able to do it tonight unfortunately so feel free to take a look if you have time!

lucacarlone replied:
@jingnanshi @pantonante : I fixed the stopping conditions and a few other things (including setting TLS to default) - please review carefully and let me know if you have more comments.

jingnanshi replied:
@lucacarlone : Had two very minor fixes. Everything looks good!

lucacarlone replied:
perfect - thanks @jingnanshi !

lucacarlone commented:
@jingnanshi @dellaert @pantonante : I fixed the outstanding issue with optimizerType. I also added a couple of unit tests (and comments) to make sure the inlier threshold is set correctly. @jingnanshi : the only thing I'm not convinced about is the stopping condition for TLS - please check my comment. Besides that, this PR is ready for review!

pantonante commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,388 @@
+  /// GNC parameters
+  BaseOptimizerParameters baseOptimizerParams; /*optimization parameters used to solve the weighted least squares problem at each GNC iteration*/
+  /// any other specific GNC parameters:
+  RobustLossType lossType = GM; /* default loss*/
```
Probably we want TLS as default loss type.

pantonante commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,425 @@
+  }
+
+  /// check for convergence between consecutive GNC iterations
+  bool checkConvergence(const double mu,
+                        const double cost,
+                        const double prev_cost) const {
+    switch (params_.lossType) {
+    case GncParameters::GM:
+      return checkMuConvergence(mu);
+    case GncParameters::TLS:
+      return checkCostConvergence(cost, prev_cost);
+    default:
+      throw std::runtime_error(
+          "GncOptimizer::checkMuConvergence: called with unknown loss type.");
+    }
+  }
```
This will work well, but it looks weird that we use two very different stopping criteria for the two robust costs. I would use the cost based convergence test in both cases as default and add a parameter to support three convergence tests:
1. cost based only (default)
2. weighs based only
3. [optional] cost based OR weights base (whatever converges first)
I think this would add clarity overall.

lucacarlone replied:
for simplicity, I would add both stopping conditions (for both costs) and give the user control of the tolerances for stopping, such that s/he can enable/disable the conditions s/he prefers.

lucacarlone replied:
@jingnanshi : let me know if you can take a look at this, otherwise I can do it later in December.

jingnanshi replied:
@lucacarlone @pantonante For the TLS case, since mu is increasing, it is difficult to check for mu (unless I have misunderstood it). What do you guys recommend?

lucacarlone replied:
@jingnanshi : Pasquale was suggesting two checks I think:
1) relative change in cost: e.g., | cost_t - cost_t-1 | /  |cost_t | < params_.relativeCostTol (then stop)
2) weights: e.g., if | weight - round(weights) | < params_.weightConvergenceTol  (then stop)
In the latter round(weights) rounds the entries of weights in {0;1} and this is mostly effective for TLS (it checks whether you have a clear decision inlier/outliers).

Note: you can also keep the stopping condition based on mu for GM, while there is no such condition for TLS

dellaert commented on gtsam/nonlinear/GaussNewtonOptimizer.h:
```diff
@@ -28,6 +28,8 @@ class GaussNewtonOptimizer;
  */
 class GTSAM_EXPORT GaussNewtonParams : public NonlinearOptimizerParams {
+public:
+  typedef GaussNewtonOptimizer OptimizerType;
```
Change to `using OptimizerType = GaussNewtonOptimizer`

dellaert commented on gtsam/nonlinear/LevenbergMarquardtParams.h:
```diff
@@ -40,6 +42,7 @@ class GTSAM_EXPORT LevenbergMarquardtParams: public NonlinearOptimizerParams {
 
   static VerbosityLM verbosityLMTranslator(const std::string &s);
   static std::string verbosityLMTranslator(VerbosityLM value);
+  typedef LevenbergMarquardtOptimizer OptimizerType;
```
Change to `using OptimizerType = LevenbergMarquardtOptimizer `

dellaert commented on tests/testGncOptimizer.cpp:
```diff
@@ -0,0 +1,670 @@
+TEST(GncOptimizer, gncParamsConstructor) {
+  // check params are correctly parsed
+  LevenbergMarquardtParams lmParams;
+  GncParams<LevenbergMarquardtParams> gncParams1(lmParams);
```
Do we need the <> here (and below) or can the type be derived from the argument? I think it can and should be omitted.

lucacarlone replied:
@dellaert : it does not work if I omit - should I do something else to make it derive the type?

dellaert replied:
Hmmm, I had expected that to work, but maybe it's only since C++17: https://en.cppreference.com/w/cpp/language/class_template_argument_deduction

dellaert commented on tests/testGncOptimizer.cpp:
```diff
@@ -0,0 +1,670 @@
+  initial.insert(X(1), p0);
+
+  LevenbergMarquardtParams lmParams;
+  GncParams<LevenbergMarquardtParams> gncParams(lmParams);
```
If default. I think we should test here the use of 
```
GncParams<LevenbergMarquardtParams> gncParams;
```
i.e. without arguments

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+namespace gtsam {
+
+/* ************************************************************************* */
+template<class BaseOptimizerParameters>
```
There is enough code here to warrant a separate file for GncParams

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+  typedef typename BaseOptimizerParameters::OptimizerType OptimizerType;
+
+  /** Verbosity levels */
+  enum VerbosityGNC {
```
GNC here is redundant

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+  };
+
+  /** Choice of robust loss function for GNC */
+  enum RobustLossType {
```
I propose to make this its own class called `GncLossType`, not an inner class. Will make things less verbose.

lucacarlone replied:
@dellaert : what do you envision we should put in the new class? this is currently just an enum

dellaert replied:
Just before the params class: otherwise you have to append the entire (templated) parameter class name which is unwieldy.

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+  }
+
+  /// GNC parameters
+  BaseOptimizerParameters baseOptimizerParams; /*optimization parameters used to solve the weighted least squares problem at each GNC iteration*/
```
Single line comments after a line should use doxygen `///< Optimization parameters...
Also, it seems this line is too long so probably you did not format with Google style? Please do so, and if comment is not single line, add it above with doxygen /** ... */

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+    SILENT = 0, SUMMARY, VALUES
+  };
+
+  /** Choice of robust loss function for GNC */
```
Single line doxygen should use `///`

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+  /// GNC parameters
+  BaseOptimizerParameters baseOptimizerParams; /*optimization parameters used to solve the weighted least squares problem at each GNC iteration*/
+  /// any other specific GNC parameters:
+  RobustLossType lossType = TLS; /* default loss*/
```
`///< Default loss.`

Note capitalization and punctuation :-)

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+  typedef typename GncParameters::OptimizerType BaseOptimizer;
+
+private:
+  NonlinearFactorGraph nfg_;
```
Document

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+  NonlinearFactorGraph getFactors() const {
+    return NonlinearFactorGraph(nfg_);
+  }
+  /// Access a copy of the internal values
```
Here and below: add return. Make sure to format with clang-format or cpp-format w Google style.

dellaert commented on gtsam/nonlinear/GncOptimizer.h:
```diff
@@ -0,0 +1,459 @@
+    }
+  }
+
+  /// update the gnc parameter mu to gradually increase nonconvexity
```
Capitalize and add period to all comments.

dellaert reviewed (commented):
Awesome! First round, will look again after comments resolved :-)

dellaert commented on gtsam/nonlinear/GncParams.h:
```diff
@@ -0,0 +1,155 @@
+  void setLossType(const GncLossType type) {
+    lossType = type;
+  }
+  /// Set the maximum number of iterations in GNC (changing the max nr of iters might lead to less accurate solutions and is not recommended).
```
Add newline between methods...

dellaert reviewed (approved):
Very cool! 
Just a few more code nits, feel free to merge after addressed ;-)

Now we need to apply it to Shonan and compare with robuts rotation averaging, right @jingwuOUO  :-) ?

lucacarlone commented:
all set! I'll merge after @pantonante and @jingnanshi  approve!

jingnanshi reviewed (approved):
:+1:
